File: sdk/reproit-android/src/test/kotlin/com/reproit/android/CaptureBatchTest.kt
package com.reproit.android

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The emitted `capture-batch-v1` for a failure occurrence that carries recorded
 * exchanges. Shape parity with `sdk/reproit-backend-node/capture.js`; the wire
 * itself is proven against the Rust protocol validator by
 * `emit_capture_batch.kt` in the host test runner.
 */
class CaptureBatchTest {

  private fun exchange(host: String, status: Int, monotonicNs: Long) =
    CapturedExchange(
      subject = host,
      exchange =
        captureHttpExchange(
          method = "GET",
          url = "https://$host/prices",
          requestHeaders = mapOf("Accept" to "application/json"),
          requestBody = null,
          status = status,
          responseHeaders = mapOf("Content-Type" to "application/json"),
          responseBody = """{"prices":null}""".toByteArray(),
        ),
      atMs = 1753747200000L,
      monotonicNs = monotonicNs,
    )

  private fun batch(exchanges: List<CapturedExchange>): Map<String, Any?>? =
    buildFailureCaptureBatch(
      appId = "com.example.app",
      sessionId = "session-android-1753747200000",
      operation = "crash:NullPointerException",
      triggerAction = "tap:key:checkout",
      signature = "crash:java.lang.NullPointerException",
      summary = "java.lang.NullPointerException: prices was null",
      observationPoint = "com.example.Checkout.total(Checkout.kt:42)",
      exchanges = exchanges,
      envelope =
        determinismEnvelope(1753747200000L, "14", "arm64-v8a", "c0ffee00c0ffee00", null),
      buildVersion = "1.2.3",
      buildCommit = "abc123",
      observedAtIso = "2026-07-30T12:00:00Z",
      batchSequence = 1,
      observedAtMs = 1753747200000L,
    )

  @Test
  fun theBatchCarriesTheCausalSequenceAndEveryExchange() {
    val built = assertNotNull(batch(listOf(exchange("pricing.internal", 200, 4_000_000L))))
    val value = batch(listOf(exchange("pricing.internal", 200, 4_000_000L)))!!
    assertEquals(1, value["version"])
    assertEquals("com.example.app", value["projectId"])
    @Suppress("UNCHECKED_CAST") val emitter = value["emitter"] as Map<String, Any?>
    assertEquals("mobile-android", emitter["id"])
    assertEquals("runtime-sdk", emitter["kind"])
    assertEquals("android", emitter["runtime"])
    @Suppress("UNCHECKED_CAST") val deployment = value["deployment"] as Map<String, Any?>
    assertEquals("1.2.3", deployment["version"])
    assertEquals("abc123", deployment["commit"])

    @Suppress("UNCHECKED_CAST") val events = value["events"] as List<Map<String, Any?>>
    val kinds =
      events.map {
        @Suppress("UNCHECKED_CAST") val event = it["event"] as Map<String, Any?>
        event["kind"]
      }
    assertEquals(
      listOf(
        "operation-start",
        "trigger",
        "checkpoint",
        "dependency",
        "operation-end",
        "observation",
      ),
      kinds,
    )
    // Dense sequence, and each event names its causal parent.
    events.forEachIndexed { index, event -> assertEquals(index + 1, event["sequence"]) }
    for (index in 1 until events.size) {
      @Suppress("UNCHECKED_CAST") val parents = events[index]["causalParentIds"] as List<String>
      assertEquals(listOf(events[index - 1]["id"]), parents)
    }
    @Suppress("UNCHECKED_CAST") val first = events[0]["causalParentIds"] as List<String>
    assertTrue("the first event has no parent", first.isEmpty())
    assertNotNull(built)
  }

  @Test
  fun theEnvelopeRidesAsANamedCheckpointAndTheExchangeIsReplayable() {
    val value = batch(listOf(exchange("pricing.internal", 200, 4_000_000L)))!!
    @Suppress("UNCHECKED_CAST") val events = value["events"] as List<Map<String, Any?>>
    @Suppress("UNCHECKED_CAST") val checkpoint = events[2]["event"] as Map<String, Any?>
    assertEquals("determinism-envelope", checkpoint["name"])
    @Suppress("UNCHECKED_CAST") val attributes = checkpoint["attributes"] as Map<String, Any?>
    assertEquals("c0ffee00c0ffee00", attributes["replaySeed"])

    @Suppress("UNCHECKED_CAST") val dependency = events[3]["event"] as Map<String, Any?>
    assertEquals("service", dependency["system"])
    assertEquals("call", dependency["operation"])
    assertEquals("pricing.internal", dependency["subject"])
    @Suppress("UNCHECKED_CAST") val captured = dependency["value"] as Map<String, Any?>
    assertEquals("replayable", captured["representation"])
    assertEquals("redacted-at-source", captured["redaction"])
    @Suppress("UNCHECKED_CAST") val payload = captured["value"] as Map<String, Any?>
    assertEquals("http", payload["protocol"])
    // Real monotonic offsets, not the ordinal fallback.
    assertEquals(4_000_000L, events[3]["monotonicNs"])
  }

  @Test
  fun theNetworkCapabilityIsClaimedOnlyWhenExchangesExist() {
    val withExchanges = batch(listOf(exchange("pricing.internal", 200, 1L)))!!
    @Suppress("UNCHECKED_CAST")
    val claimed = withExchanges["capabilities"] as List<Map<String, Any?>>
    assertTrue(claimed.any { it["capability"] == "network" && it["completeness"] == "complete" })

    val without = batch(emptyList())!!
    @Suppress("UNCHECKED_CAST") val bare = without["capabilities"] as List<Map<String, Any?>>
    assertFalse(
      "an empty capture must never claim complete network coverage",
      bare.any { it["capability"] == "network" },
    )
  }

  @Test
  fun anUnusableProjectIdentifierRefusesToBuildABatch() {
    // The ingest protocol rejects these, so the SDK keeps the legacy path
    // rather than emitting something the server will drop.
    val bad =
      buildFailureCaptureBatch(
        appId = "not a token",
        sessionId = "session-1",
        operation = "crash",
        triggerAction = "load",
        signature = "crash:X",
        summary = "boom",
        observationPoint = "X",
        exchanges = emptyList(),
        envelope = determinismEnvelope(1L, "14", "arm64", "00ff00ff00ff00ff", null),
        buildVersion = null,
        buildCommit = null,
        observedAtIso = "2026-07-30T12:00:00Z",
        batchSequence = 1,
        observedAtMs = 1L,
      )
    assertNull(bad)
    assertTrue(validToken("com.example.app"))
    assertFalse(validToken("not a token"))
  }

  @Test
  fun anAbsentBuildIdentityOmitsDeploymentEntirely() {
    val value =
      buildFailureCaptureBatch(
        appId = "com.example.app",
        sessionId = "session-1",
        operation = "crash",
        triggerAction = "load",
        signature = "crash:X",
        summary = "boom",
        observationPoint = "X",
        exchanges = emptyList(),
        envelope = determinismEnvelope(1L, "14", "arm64", "00ff00ff00ff00ff", null),
        buildVersion = null,
        buildCommit = null,
        observedAtIso = "2026-07-30T12:00:00Z",
        batchSequence = 1,
        observedAtMs = 1L,
      )!!
    assertNull(value["deployment"])
  }
}

File: sdk/reproit-react-native/test/capture-batch.test.ts
import { execFileSync } from 'node:child_process';
import { resolve } from 'node:path';

import { buildCaptureBatch, buildEnvelope, replaySeed, validToken } from '../src/capture-batch';
import { buildHttpExchange } from '../src/exchange';

const WORKSPACE = resolve(__dirname, '../../..');

function sampleBatch() {
  const exchange = buildHttpExchange(
    {
      method: 'GET',
      url: 'https://api.test/prices?tier=gold',
      headers: { 'content-type': 'application/json' },
      body: null,
    },
    {
      status: 200,
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ prices: null }),
    },
  );
  return buildCaptureBatch({
    appId: 'app-demo',
    sessionId: 'rn-1753747200000',
    batchId: 'cb-rn-1753747200000-1',
    deployment: { version: '1.4.2', commit: 'abc123' },
    observedAt: '2026-07-27T12:00:00Z',
    occurrence: {
      operation: 'tap:key:testid:quote',
      trigger: { path: [{ sig: 'a1b2c3d4', action: 'tap:key:testid:quote' }] },
      exchanges: [{ ...exchange, at: 1753747200009, monoNs: 9000000 }],
      failure: {
        oracle: 'crash',
        summary: "TypeError: Cannot read property '0' of null",
        signature: 'crash:a1b2c3d4',
        observationPoint: 'QuoteScreen',
      },
      envelope: buildEnvelope({
        observedAtMs: 1753747200000,
        platform: 'ios',
        osVersion: '18.2',
        locale: 'en-US',
        timezone: 'Europe/Berlin',
        replaySeed: 'c0ffee00c0ffee00',
      }),
    },
  });
}

describe('capture batch emission', () => {
  test('the emitted batch passes the protocol validator', () => {
    const batch = sampleBatch();
    const output = execFileSync(
      'cargo',
      ['run', '-q', '-p', 'reproit-protocol', '--bin', 'capture-validate'],
      { cwd: WORKSPACE, input: JSON.stringify(batch), encoding: 'utf8' },
    );
    expect(output).toContain('portable');
  }, 300_000);

  test('the batch carries the trigger, exchanges, envelope, and observation in order', () => {
    const kinds = sampleBatch().events.map((event) => event.event.kind);
    expect(kinds).toEqual([
      'operation-start',
      'trigger',
      'checkpoint',
      'dependency',
      'operation-end',
      'observation',
    ]);
  });

  test('the dependency event nests the raw exchange for the replay projection', () => {
    const dependency = sampleBatch().events.find((event) => event.event.kind === 'dependency');
    const value = dependency?.event.value as { representation: string; value: Record<string, unknown> };
    expect(value.representation).toBe('replayable');
    expect(value.value.kind).toBe('effect');
    const exchange = value.value.exchange as { response: { body: unknown; status: number } };
    expect(exchange.response.status).toBe(200);
    expect(exchange.response.body).toEqual({ prices: null });
    // Real monotonic offsets ride from the recorded exchange, not the ordinal.
    expect(dependency?.monotonicNs).toBe(9000000);
  });

  test('the determinism envelope states what a device can honestly know', () => {
    const checkpoint = sampleBatch().events.find((event) => event.event.kind === 'checkpoint');
    expect(checkpoint?.event.name).toBe('determinism-envelope');
    const attributes = checkpoint?.event.attributes as Record<string, unknown>;
    expect(attributes.runtime).toBe('react-native');
    expect(attributes.tz).toBe('Europe/Berlin');
    expect(attributes.os).toBe('ios');
    expect(attributes.replaySeed).toBe('c0ffee00c0ffee00');
    expect(attributes.observedAtMs).toBe(1753747200000);
    // A device has no process arch or image digest; those are omitted, not guessed.
    expect(attributes.arch).toBeUndefined();
    expect(attributes.imageDigest).toBeUndefined();
  });

  test('deployment identity rides when supplied and is omitted when absent', () => {
    expect(sampleBatch().deployment).toEqual({ version: '1.4.2', commit: 'abc123' });
    const batch = buildCaptureBatch({
      appId: 'app-demo',
      sessionId: 'rn-1',
      batchId: 'cb-rn-1',
      deployment: null,
      occurrence: {
        operation: 'load',
        trigger: null,
        exchanges: [],
        failure: {
          oracle: 'crash',
          summary: 'boom',
          signature: 'crash:x',
          observationPoint: 'load',
        },
        envelope: buildEnvelope({ observedAtMs: 1, replaySeed: replaySeed() }),
      },
    });
    expect(batch.deployment).toBeUndefined();
    expect(validToken(batch.batchId)).toBe(true);
    expect(validToken(batch.projectId)).toBe(true);
  });

  test('the replay seed is a bounded hex token', () => {
    expect(replaySeed()).toMatch(/^[0-9a-f]{16}$/);
  });
});
